pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -36,152 +39,218 @@ static int make_amp_array(const int arr_rows, const int arr_cols, SingleGroup *
-  
-  /* array for amp data and amp data + gap pixels */
-  double * ampdata, * ampdata_gap;
-  
+
   const double serial_freq = 1000./22.;    /* serial pixel frequency */
   const double parallel_shift = 3.212;     /* parallel shift time */
   
   /* number of virtual pixels at end of each row */
   const int ngap_pix = (int) (serial_freq * parallel_shift + 0.5);
   
-  const int arr_rows = chip2->sci.data.ny;
-  const int arr_cols = chip2->sci.data.nx/2;
+  /* There is always at least one SingleGroup - use sg[0] to get the image size of a single amp.
+   * If this is a full frame image, then "nx" represents all the columns across both amps plus
+   * the physical overscan and must be divided by 2.  If this is a subarray, then "nx" represents
+   * just the columns for a single amp plus physical overscan.
+   */
+  const int arr_rows = sg[0]->sci.data.ny;
+  const int arr_cols = (acs->subarray == NO) ? sg[0]->sci.data.nx/2 : sg[0]->sci.data.nx;
   
   /* total number of real and gap pixels per quadrant */
   const int nquad_pix = (arr_cols + ngap_pix) * arr_rows;
   
-  /* array of true DC bias levels */
-  double * dc_bias_levels;
-  
   /* arrays below are in order of amp A, B, C, D */
   
-  /* time constant of AC high pass filter in external pre-amp. */
+  /* time constant of AC high pass filter in external pre-amp in microseconds */
   const double time_const[NAMPS] = {37.290251, 36.180001, 37.867770, 42.461249};
 
   /* ratio of DC offset shift and pixel signal */
   const double dc_ratio[NAMPS] = {0.3, 0.3, 0.3, 0.3};
   
   /* DSI sensitivity */
   const double dsi_sens[NAMPS] = {2.9188919e-3, 10.805754e-3, 12.432145e-3, 3.8596253e-3};
-    
-  /* factor combining time constant and clocking frequency */
-  double factor;
-  
-  /* summation variables */
-  double sum;
-  int num;
-  double magic_square_mean;
-  
-  /* allocate space for data arrays */
-  ampdata = malloc(arr_rows * arr_cols * sizeof(double));
-  ampdata_gap = malloc(nquad_pix * sizeof(double));
-  
-  /* allocate space for true DC bias levels array */
+
+  PtrRegister ptrReg;
+  initPtrRegister(&ptrReg);
+
+  /* array of true DC bias levels */
+  double * dc_bias_levels = NULL;
   dc_bias_levels = malloc((nquad_pix + 1) * sizeof(double));
-  
-  for (i = 0; i < NAMPS; i++) {
-    /* put a single amp's data into ampdata */
-    if (i < 2) {
-      make_amp_array(arr_rows, arr_cols, chip1, i, ampdata);
-    } else {
-      make_amp_array(arr_rows, arr_cols, chip2, i, ampdata);
-    }
-       
-    /* calculate "magic square mean" */
-    sum = 0.0;
-    num = 0;
+  addPtr(&ptrReg, dc_bias_levels, &free);
     
-    for (j = 2057; j <= 2066; j++) {
-      for (k = 13; k <= 22; k++) {
-        sum += ampdata[arr_cols*j + k];
-        num++;
+  /* array for amp data and amp data + gap pixels */
+  double *ampdata = malloc(arr_rows * arr_cols * sizeof(double));
+  addPtr(&ptrReg, ampdata, &free);
+  double *ampdata_gap = malloc(nquad_pix * sizeof(double));
```

Comment:
All of the above ``malloc``s need to be ``NULL`` checked handling ``OUT_OF_MEMORY`` errors.

Reply:
Indeed.  Went to great trouble to add these checks in other routines, but I neglected to add the checks here.